### PR TITLE
Fix a buffer overflow in hmac_setup_internal

### LINKED ISSUE #254: Buffer overflow in hmac setup on known unsupported algorithm, with long input
### Description

When calling `psa_mac_sign_setup()` for HMAC with a known but unsupported algorithm and a key that's longer than the block size, a buffer overflow occurs.
This happens in `psa_hmac_setup_internal()`, if `psa_hash_setup() or `psa_hash_update()` fail, and the `key_length` is larger than PSA_HMAC_MAX_HASH_BLOCK_SIZE, because at the cleanup we do the following:
~~~
mbedtls_platform_zeroize( ipad, key_length );
~~~

### Issue request type

    [ ] Question
    [ ] Enhancement
    [x ] Bug

### PR DESCRIPTION
**Pull request by RonEld**
At the end of `psa_hmac_setup_internal()`, the ipad is cleared.
However, the size that was given to clear was `key_len` which is larger
than the size of `ipad`.
Fixes #254

### CODE REVIEW

File: library/psa_crypto.c [hunk 1]
```diff
@@ -2735,7 +2735,7 @@ static psa_status_t psa_hmac_setup_internal( psa_hmac_internal_data *hmac,
     status = psa_hash_update( &hmac->hash_ctx, ipad, block_size );
 
 cleanup:
-    mbedtls_platform_zeroize( ipad, key_length );
+    mbedtls_platform_zeroize( ipad, sizeof(ipad) );
```
Comment (Patater) [inline library/psa_crypto.c]: Do we need to fix the memcpy above as well? It also copies to ipad from key, using key_length number of bytes.

Also, how was this overflow caught? Do we need to add any regression test?

Comment (gilles-peskine-arm) [reply library/psa_crypto.c]: The memcpy call is only done if `key_length <= block_size` and the function checks that `block_size <= sizeof( ipad )` earlier. So _that_'s ok. However, there are other uses of `key_length` that aren't ok.

Comment (RonEld) [reply library/psa_crypto.c]: I caught this using the `psa_crypto_test` on  target that doesn't have support for SHA384 (NRF52480). `psa_hash_setup()` returned an error, going to cleanup.
The `key_length` was 131, and `PSA_HMAC_MAX_HASH_BLOCK_SIZE` is 128

**Comment by RonEld** [comment]
>And please add at least one test case with a key that's longer than the block size. Evidently we're missing that.

I am not sure we are missing this, because calling `psa_hash_finish()` will update `key_length` to valid size.
I believe this issue will happen *only* when `psa_hash_setup` or `psa_hash_update` fail.

> think a better fix would be to set key_length = hash_size inside the if( key_length > block_size ) block, and make no other change.

I think this will change hte algorithm, as `psa_hash_update()` uses the original key length to hash the key. After this call, `psa_hash_finish()` should already update `key_length` to the hash length.  As mentioned in previous sentence, only if calls to `psa_hash_setup()` and `psa_hash_update()` fail, then we encounter this overflow, which means updating the `key_length` before every `goto cleanup` which I think is not as clean as updating this line alone.

As for adding regression tests, I can't think of a test without fuzzing

**Comment by gilles-peskine-arm** [comment]
For the non-regression testing, I'd expect at least Asan or Msan to detect corruption if there's a buffer overflow. Maybe we need a test with a slightly longer key. Or maybe what we're missing is a test that tries to do HMAC with a hash algorithm that isn't supported _and_ a long key.

> After this call, psa_hash_finish() should already update key_length to the hash length.

Oh, right. So the only incorrect case was if `psa_hash_setup` or `psa_hash_update` inside the `if( key_length > block_size )` fails. So your fix is correct, but please do add a test.

**Comment by Patater** [comment]
We could benefit from a test that mocks `psa_hash_setup` or `psa_hash_update` to return errors at certain points. This would help increase our branch coverage.

In this particular case, there may be another way to get equivalent branch coverage.

**Comment by RonEld** [comment]
The CI failures in TLS tests don't seem related.
Both are:
~~~
DTLS client reconnect from same port: reconnect ........................ FAIL
  ! pattern 'resend' MUST NOT be present in the Client output
  ! outputs saved to o-XXX-602.log
~~~

**Comment by RonEld** [comment]
>In this particular case, there may be another way to get equivalent branch coverage.

I am open for suggestions

**Comment by RonEld** [comment]
>For the non-regression testing, I'd expect at least Asan or Msan to detect corruption if there's a buffer overflow. Maybe we need a test with a slightly longer key. Or maybe what we're missing is a test that tries to do HMAC with a hash algorithm that isn't supported and a long key.

I don't think this would work. A longer key will just have longer input for `psa_hash_update()`. An algorithm that isn't supported, will just skip the test itself, becasue the `MBEDTLS_SHAXXX_C` is not defined. The only thing I can think f is a mocktest, as suggested in https://github.com/ARMmbed/mbed-crypto/pull/248#issuecomment-529964298

**Comment by gilles-peskine-arm** [comment]
I was thinking of something like
```
PSA MAC setup: bad algorithm (unknown MAC algorithm)
depends_on:!MBEDTLS_MD5_C
mac_setup:PSA_KEY_TYPE_HMAC:"a 33-byte key, or longer if that's what it takes":PSA_ALG_HMAC(MBEDTLS_MD5_C):PSA_ERROR_NOT_SUPPORTED
```
Expected result before your fix: crash under Asan. Expected result after your fix: pass.

**Comment by RonEld** [comment]
I'll check this

**Comment by Patater** [comment]
> > In this particular case, there may be another way to get equivalent branch coverage.
> 
> I am open for suggestions

Could you try with psa_mac_setup and PSA_ALG_HMAC(PSA_ALG_SHA3_224)? We don't have SHA3 yet. I'm not quite sure how to make key size larger than hash block size.

**Comment by Patater** [comment]
> depends_on:!MBEDTLS_MD5_C

Would this test ever run? We pretty much always have MD5 enabled, don't we?

**Comment by Patater** [comment]
Hmm, `psa_get_hash_block_size()` returns `0` for unsupported hashes. That would mean any key length is longer than any unsupported hash's block size.

**Comment by gilles-peskine-arm** [comment]
> Would this test ever run? We pretty much always have MD5 enabled, don't we?

For every hash *h*, we run the test suite at least once with *h* disabled, via `depends-hashes.pl`.

**Comment by Patater** [comment]
> > Would this test ever run? We pretty much always have MD5 enabled, don't we?
> 
> For every hash _h_, we run the test suite at least once with _h_ disabled, via `depends-hashes.pl`.

Great. MD5 is a good choice then. psa_get_hash_block_size() will tell you the hash block size even when MD5 is disabled, which is what we want to get into this branch.

**Comment by RonEld** [comment]
I added a test.
Without the fix:
~~~
test_suite_poly1305 ............................................... PASS
test_suite_psa_crypto ............................................. *** stack smashing detected ***: ./test_suite_psa_crypto terminated
FAIL
Use of uninitialized value $skipped in subtraction (-) at scripts/run-test-suites.pl line 114.
Use of uninitialized value $tests in subtraction (-) at scripts/run-test-suites.pl line 114.
test_suite_psa_crypto_entropy ..................................... PASS
~~~

With the fix:
~~~
test_suite_poly1305 ............................................... PASS
test_suite_psa_crypto ............................................. PASS
test_suite_psa_crypto_entropy ..................................... PASS
~~~

**Comment by Patater** [comment]
CI failures in merge job is only stale generated files, which will get updated when Mbed TLS updates the submodule.

File: tests/suites/test_suite_psa_crypto.data [hunk 2]
```diff
@@ -995,6 +995,10 @@ PSA MAC verify: CMAC-AES-128, truncated to 4 bytes
 depends_on:MBEDTLS_CMAC_C:MBEDTLS_AES_C
 mac_verify:PSA_KEY_TYPE_AES:"2b7e151628aed2a6abf7158809cf4f3c":PSA_ALG_TRUNCATED_MAC(PSA_ALG_CMAC, 4):"6bc1bee22e409f96e93d7e117393172aae2d8a571e03ac9c9eb76fac45af8e5130c81c46a35ce411":"dfa66747"
 
+PSA MAC setup: incompatible key MD for HMAC
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto.data]: It's not-supported, not incompatible. And it's relevant that this is with a long key (longer than the block size).

> PSA MAC setup: algorithm known but not supported, long key

Would be good to also have a test case with a short key (less than the block size), for completeness.

**Comment by RonEld** [comment]
@gilles-peskine-arm done

**Comment by RonEld** [comment]
created https://github.com/ARMmbed/mbed-crypto/issues/254 to track this

**Comment by Patater** [comment]
CI failures are stale generated files (to be fixed when Mbed TLS moves to new Mbed Crypto version) and the following known-flaky test.

```
DTLS client reconnect from same port: reconnect ........................ FAIL
  ! pattern 'resend' MUST NOT be present in the Client output
  ! outputs saved to o-XXX-602.log
```

**Pull request merged**